### DISCUSSION TITLE
Validating query parameters

### DISCUSSION THREAD

**Discussion by ad-on-is** [adonisjs/core#3519] [General]
I'm using `uuid_generate_v4()`, and when I pass some invalid format via query, I get an error, I'd rather want to hide from the end-user.

`GET /users/1` throws an error like this

```
"select * from \"users\" where \"id\" = $1 and \"users\".\"deleted_at\" is null limit $2 - invalid input syntax for type uuid: \"1\"",
```

I'm aware the validator can be used standalone, but I'd need to do this in each and every Controller. Is there a simpler solution?

**Comment by Julien-R44** [comment]
You could use Params matchers : https://docs.adonisjs.com/guides/routing#params-matchers
There is even a builtin matcher to compare uuids: https://docs.adonisjs.com/guides/routing#inbuilt-matchers

In this case if the uuid sent is not valid, then the router will not redirect the action to your controller and your application will return a 404. 

Example : 
```ts
Route
  .get('/users/:id', 'MyController.show')
  .where('id', Route.matchers.uuid())
```

**Reply by ad-on-is** [reply]
Cool... that sounds great. I didn't even know about that. 👍

But...my only problem is, I've built all my routes with `Route.resource()` and something like `Route.resource('/users', 'Controller').where('id', Route.matchers.uuid())` does not work.

In the meantime, I've implemented the standalone validator in a middleware, to check if `ctx.request.qs()` and `ctx.request.params()` contain anything that has `id` or `_id` in it, and validates it with a `rules.uuid({version: 4})`

**Reply by ad-on-is** [reply]
Lol... I even went that far to try this out, without luck... but maybe It could be implemented in future releases:

```js
Route.resource('users', 'UsersController')
          .apiOnly()
          .where({
            show: { id: Route.matchers.uuid() },
          })
```

**Reply by Julien-R44** [reply]
I tried like you but using Route.matchers.number instead of Route.matchers.uuid and it works perfectly so I think the problem is coming from the regexp used to match the uuid : 
https://github.com/adonisjs/http-server/blob/eb89bd3d2b28b531c68a382ad6148807e8852dd0/src/Router/Matchers.ts#L34

I tried to match a v4 uuid with the regex in question and it doesn't work. I can't really say why without investigating, but what you can do is develop your own matcher : https://docs.adonisjs.com/guides/routing#params-matchers

And use this regex that seems to work with uuid v4 : https://stackoverflow.com/questions/7905929/how-to-test-valid-uuid-guid

EDIT : Sorry, forget what I said, it works with the current regex, I must have made a mistake somewhere ! Gonna investigate a bit later

**Reply by Julien-R44** [reply]
So actually it doesn't work if your uuid has lowercase letters (look at the regex, a uuid is considered valid only with uppercase letters), I think your problem is here. I have set up my router like this: 

```
Route.group(() => {
  Route.resource('/users', 'AccountsController').apiOnly().where('id', Route.matchers.uuid())
})
```

GET /users/4BDC681B-FD15-4257-9B23-55B69F418B48 works ( Route is matched and Controller method is called )
GET /users/96324739-0614-4654-96ef-3930db48318f doesn't works ( Route isn't matched and returns 404 )

So yes, if you need to support lowercase letters, a solution could be to write your own matcher as I suggested above

**Reply by ad-on-is** [reply]
Thank you so much for your investigation 👍 

I'll then ditch my middleware and use the custom matcher instead. This way it's more reproducible what's going on.
